Telemetry: More JAX-RS tests and bug fixes

Azquelt opened the pull request:
New Telemetry JAX-RS tests: (all for #23448)
- Test telemetry with JAX-RS server-side async (`CompletionStage` and `@Suspended AsyncResponse`)
- Test telemetry with all HTTP methods
- Test telemetry with a selection of different HTTP response codes

Bug fixes:
- Close the scope correctly for `SERVER` spans when JAX-RS server-side async is used (product bug) (fixes #24174)
- Wrap rest async client tasks correctly when `concurrent-3.0` is enabled (product bug in restfulWS) (fixes #24235)
- Don't rely on timestamps to order spans in `InMemorySpanExporter` (test bug) (fixes #24194)

Other improvements:
- Add `SpanDataMatcher` (based on existing `SpanMatcher` for Jaeger spans) to create better error messages when spans are not as expected
- Add `TestSpans` utility to create and close test spans to reduce test boilerplate.

Azquelt committed:
mpTelemetry: Test JAX-RS async resource method

Test MP Telemetry JAX-RS integration when using an async resource
method.

We already have tests for async JAX-RS and MP Rest clients.

Azquelt committed:
Close Telemetry span in Servlet listener

Use a ServletRequestListener to close the Telemetry Span, rather than
doing it in a JAX-RS response filter.

When calling async resource methods, request and response JAX-RS filters
can be called on different threads whereas the context scope must be set
and removed on the same thread.

Using a ServletRequestListener allows us to listen for when processing
is finished on the request thread, even if the response to the request
hasn't been sent yet.

Azquelt committed:
Fix Telemetry InMemorySpanExporter ordering

The timestamps on spans are unreliable for ordering the spans, at least
when a set of spans covers several requests.

Make the following changes:
 - order spans using their parent links. This is sufficient for most
   tests while more complicated scenarios will need more complex logic.
 - key span retrieval on a traceId to ensure we only look for spans for
   the current test. This removes the need to call reset() and wait
   between every test.
 - add a shutdown check for any spans which were exported but not
   retrieved by a test. This will include any spans which arrived late
   or any spans with a traceId which was not queried for.
 - add a method to ignore a traceId for the shutdown check - e.g.
   because we know a test has failed so it's likely that there will be
   traces left over at shutdown.

Azquelt committed:
Telemetry: test JAX-RS methods and status codes

Add tests for a variety of HTTP methods and status codes.

Azquelt committed:
Rest: Wrap tasks with context in supplyAsync

Update AsyncClientExecutorService so that it always applies context
wrappers to tasks passed to supplyAsync.

Previously it was not wrapping tasks when the concurrent-3.0 feature was
enabled.

Azquelt committed:
Telemetry: update jaxrs client task wrapper

Implement the new method on the ClientAsyncTaskWrapper to wrap a
Supplier.

Azquelt committed:
Update old tests for InMemorySpanExporter changes

Update old, non-jaxrs, tests to call the new InMemorySpanExporter
methods and use new TestSpans and SpanDataMatcher where appropriate.

Azquelt committed:
Telemetry: update JAX-RS tests

Several improvements to the JAX-RS Telemetry tests:
 - use the new methods on InMemorySpanExporter to make tests more
   reliable
 - Ensure that the JaxRsIntegration tests which don't use FATServlet
   still propagate any exception back to the test framework
 - Use TestSpans and SpanDataMatcher where appropriate to give better
   error messages

Azquelt committed:
Test Telemetry with JAX-RS Async Suspended method

Test Telemetry with a JAX-RS resource method which uses @Suspended
AsyncResponse. This is an alternative to returning CompletionStage which
we already had a test for.

Azquelt committed:
Telemetry: add Matcher for SpanData objects

This will improve failure messages when tests fail.

A lot of this code is taken from the existing matchers for spans
retrieved from Jaeger and Zipkin. I've pulled out the common parts into
an abstract class and plan to make the Jager and Zipkin matchers extend
from it later.

Azquelt committed:
Telemetry: capture common boilerplate pattern

Create a TestSpans helper bean for creating and closing test spans to
remove common test boilerplate.

Azquelt commented:
#libby
#build

Yas-Franco commented on dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryClientAsyncTaskWrapper.java:
```diff
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
```
The date in this header needs updating

Azquelt replied:
It has been, the latest version reads `2022, 2023`

jim-krueger commented on dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryClientAsyncTaskWrapper.java:
```diff
@@ -4,7 +4,7 @@
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
```
Add 2023

Azquelt replied:
bleh, thanks

Azquelt commented:
Rebased on integration to fix a soft merge conflict with `TelemetryLongRunningTest` and have made fixes suggested in review comments.

Azquelt commented:
Squashed review changes into their appropriate commits. (Original changes were f037d84 and fa7b7df)

Yas-Franco reviewed (dismissed):
Looks good

Azquelt commented:
#build
#libby

Azquelt commented:
Found a mistake in the code where I moved the closing of the scope to a servlet listener which caused the agent tests to fail. Now fixed.

#libby
#build

Yas-Franco reviewed (approved):
Looks good
